sweepbatcher/sweep_batch.go

```diff
@@ -1814,41 +1952,92 @@ func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
 		b.Warnf("transaction %v has no outputs", txHash)
 	}
 
+	// Determine if we should use presigned mode for the batch.
+	presigned, err := b.isPresigned()
+	if err != nil {
+		return fmt.Errorf("failed to determine if the batch %d uses "+
+			"presigned mode: %w", b.id, err)
+	}
+
+	// Sort sweeps by the addition order if this is presigned mode.
+	var allSweeps []sweep
+	if presigned {
+		allSweeps, err = b.getOrderedSweeps(ctx)
+		if err != nil {
+			return fmt.Errorf("getOrderedSweeps(%d) failed: %w",
+				b.id, err)
+		}
+	} else {
+		allSweeps = make([]sweep, 0, len(b.sweeps))
+		for _, s := range b.sweeps {
+			allSweeps = append(allSweeps, s)
+		}
+	}
+
 	// As a previous version of the batch transaction may get confirmed,
 	// which does not contain the latest sweeps, we need to detect the
 	// sweeps that did not make it to the confirmed transaction and feed
 	// them back to the batcher. This will ensure that the sweeps will enter
 	// a new batch instead of remaining dangling.
-	var totalSweptAmt btcutil.Amount
-	for _, sweep := range b.sweeps {
+	var (
+		totalSweptAmt   btcutil.Amount
+		confirmedSweeps = []wire.OutPoint{}
+		purgedSweeps    = []wire.OutPoint{}
+		purgedSwaps     = []lntypes.Hash{}
+	)
+	for _, sweep := range allSweeps {
 		found := false
 
 		for _, txIn := range spendTx.TxIn {
 			if txIn.PreviousOutPoint == sweep.outpoint {
 				found = true
 				totalSweptAmt += sweep.value
 				notifyList = append(notifyList, sweep)
+				confirmedSweeps = append(
+					confirmedSweeps, sweep.outpoint,
+				)
 			}
 		}
 
 		// If the sweep's outpoint was not found in the transaction's
 		// inputs this means it was left out. So we delete it from this
 		// batch and feed it back to the batcher.
-		if !found {
-			newSweep := sweep
-			delete(b.sweeps, sweep.outpoint)
+		if found {
+			continue
+		}
+
+		newSweep := sweep
+		delete(b.sweeps, sweep.outpoint)
+
+		newInput := Input{
+			Outpoint: newSweep.outpoint,
+			Value:    newSweep.value,
+		}
+
+		// In presigned mode we should form a SweepRequest per swap
+		// (i.e. per group) and keep them ordered.
+		L := len(purgeList)
+		if presigned && L != 0 &&
+			purgeList[L-1].SwapHash == newSweep.swapHash {
+
+			// Add the input to existing SweepRequest for this swap.
+			purgeList[L-1].Inputs = append(
+				purgeList[L-1].Inputs, newInput,
+			)
+		} else {
 			purgeList = append(purgeList, SweepRequest{
 				SwapHash: newSweep.swapHash,
-				Inputs: []Input{
-					{
-						Outpoint: newSweep.outpoint,
-						Value:    newSweep.value,
-					},
-				},
+				Inputs:   []Input{newInput},
```

Comment:
so when we are in `presigned` mode and the purged list is empty we will land here the first time, right.

Reply:
We can reach this line either in regular mode or in presigned mode in the beginning or when new swap starts.

I added a comment:

```go
} else {
		// Add the current sweep as a new element to purgeList.
		// This is possible either in regular mode or in
		// presigned mode in the beginning or on new swap.
		purgeList = append(purgeList, SweepRequest{
				SwapHash: newSweep.swapHash,
				Inputs:   []Input{newInput},
				Notifier: newSweep.notifier,
		})
}
```